transport_nantes/mobilito/views.py

```diff
@@ -29,6 +29,111 @@
+    who have only seen some tutorial records.  They depend on the
+    short timeout cookie.  Once they have seen all pages, we persist
+    the fact that they are done.
+
+    Similarly, unauthenticated users who don't authenticate will time
+    out their tutorial views with the cookie and so have to review the
+    tutorial.
+
+    The tutorial is intended to be short enough that none of this
+    should be particularly bothersome.  Our goal is simply to make
+    sure everyone who records has some chance of knowing what our
+    expectations are, what the different buttons mean.
+
+    """
+    tutorial_cookie_name = 'mobilito_tutorial'
+    all_tutorial_pages = set(['presentation', 'pietons', 'voitures',
+                              'velos', 'transports-collectifs',])
+
+    def default_page(self) -> str:
+        """Return the tutorial page to view if no page requested."""
+        return "presentation"
+
+    def canonical_page(self, page_name) -> str:
+        """Return a valid tutorial page name."""
+        if page_name in self.all_tutorial_pages:
+            return page_name
+        return self.default_page()
+
+    def persist_pages_seen(self, response, pages_seen) -> None:
+        """Persist the tutorial cookie."""
+        response.set_cookie(
+            key=self.tutorial_cookie_name,
+            value=b64encode(pickle.dumps(pages_seen)).hex(),
+            expires=datetime.now(timezone.utc) + timedelta(hours=2))
+
+    def pages_seen(self, request) -> set:
+        """Return the set of pages the user has seen.
+
+        The set is based on the tutorial cookie, and so expires when
+        it does.
+
+        """
+        if 'user' in request and request.user.is_authenticated:
+            mobilito_user = get_MobilitoUser(request)
+            if mobilito_user.completed_tutorial_timestamp:
+                return all_tutorial_pages
+        tutorial_cookie = request.COOKIES.get(
+            self.tutorial_cookie_name, '')
+        if not tutorial_cookie:
+            return set()
+        pages_seen: set = pickle.loads(b64decode(bytes.fromhex(tutorial_cookie)))
+        return pages_seen
+
+    def pages_to_see(self, request, this_page=None) -> set:
+        """Return the set of pages the user must see.
+
+        If the user is still required to view tutorial pages, return
+        the set required.  If we are composing a tutorial page for the
+        user, do not count it as needing to be seen, since it will
+        have been see by the time it matters.
+
+        """
+        if 'user' in request and request.user.is_authenticated:
+            mobilito_user = get_MobilitoUser(request)
+            if mobilito_user.completed_tutorial_timestamp:
+                # User has seen all tutorial pages, nothing left to see.
+                return set()
+        return self.all_tutorial_pages - self.pages_seen(request) - set([this_page])
+
+    def next_page_to_see(self, request, this_page) -> str:
```

Comment:
The return type hinting is missing the "None", you could use Union[str, None] to tell that the function may return None as well, not only strings